File: samples/java/SupportChat/Server/Support/src/main/java/systems/zlink/samples/supportchat/server/support/spots/entryspot/handlers/OpenConversationActorHandler.java
package systems.zlink.samples.supportchat.server.support.spots.entryspot.handlers;

import systems.zlink.framework.ZLinkMessageContext;
import systems.zlink.framework.spots.ZLinkEntrySpotActorRequestHandler;
import systems.zlink.samples.supportchat.server.configuration.SampleNames;
import systems.zlink.samples.supportchat.server.configuration.SampleTimings;
import systems.zlink.samples.supportchat.server.support.actors.SupportUserActor;
import systems.zlink.samples.supportchat.server.support.spots.entryspot.SupportEntrySpot;
import systems.zlink.samples.supportchat.shared.contracts.Messages;

import java.util.concurrent.CompletionStage;

public final class OpenConversationActorHandler
        implements ZLinkEntrySpotActorRequestHandler<
                SupportEntrySpot,
                SupportUserActor,
                Messages.OpenConversationReq,
                Messages.OpenConversationRes> {
    @Override
    public CompletionStage<Messages.OpenConversationRes> handle(
            SupportEntrySpot spot,
            SupportUserActor actor,
            ZLinkMessageContext context,
            Messages.OpenConversationReq request) {
        requireRole(actor, SampleNames.Roles.Customer);
        // --8<-- [start:doc-sc-open-actor]
        return spot.context()
                .outbound()
                .requestToChannel(
                        SampleNames.ApiChannel,
                        new Messages.OpenConversationApiReq(
                                actor.actorId(), actor.displayName(), request.subject()))
                .timeout(SampleTimings.RequestTimeout)
                .submit(Messages.OpenConversationApiRes.class)
                .thenApply(
                        opened -> {
                            Messages.JoinConversationRes scheduled =
                                    actor.scheduleConversationJoin(
                                            opened.conversationId(),
                                            request.subject(),
                                            new Messages.JoinConversationReq(
                                                    opened.conversationId(),
                                                    actor.participantId(),
                                                    actor.role(),
                                                    actor.displayName()));
                            return new Messages.OpenConversationRes(
                                    opened.conversationId(), scheduled.state());
                        });
        // --8<-- [end:doc-sc-open-actor]
    }

    private static void requireRole(SupportUserActor actor, String expectedRole) {
        if (!expectedRole.equals(actor.role())) {
            throw new IllegalStateException(
                    "Expected role " + expectedRole + " but got " + actor.role());
        }
    }
}

File: samples/kotlin/SupportChat/Server/Support/src/main/kotlin/systems/zlink/samples/kotlin/supportchat/server/support/infrastructure/zlink/spots/entryspot/handlers/OpenConversationActorHandler.kt
package systems.zlink.samples.kotlin.supportchat.server.support.infrastructure.zlink.spots.entryspot.handlers

import systems.zlink.framework.ZLinkMessageContext
import systems.zlink.framework.kotlin.ZLinkSuspendingEntrySpotActorRequestHandler
import systems.zlink.framework.kotlin.await
import systems.zlink.samples.kotlin.supportchat.server.configuration.SampleNames
import systems.zlink.samples.kotlin.supportchat.server.configuration.SampleTimings
import systems.zlink.samples.kotlin.supportchat.server.configuration.SupportChatRoles
import systems.zlink.samples.kotlin.supportchat.server.support.infrastructure.zlink.actors.SupportUserActor
import systems.zlink.samples.kotlin.supportchat.server.support.infrastructure.zlink.spots.entryspot.SupportEntrySpot
import systems.zlink.samples.kotlin.supportchat.shared.contracts.JoinConversationReq
import systems.zlink.samples.kotlin.supportchat.shared.contracts.OpenConversationApiReq
import systems.zlink.samples.kotlin.supportchat.shared.contracts.OpenConversationApiRes
import systems.zlink.samples.kotlin.supportchat.shared.contracts.OpenConversationReq
import systems.zlink.samples.kotlin.supportchat.shared.contracts.OpenConversationRes

class OpenConversationActorHandler :
    ZLinkSuspendingEntrySpotActorRequestHandler<
        SupportEntrySpot,
        SupportUserActor,
        OpenConversationReq,
        OpenConversationRes,
    > {
    override suspend fun handle(
        entrySpot: SupportEntrySpot,
        actor: SupportUserActor,
        context: ZLinkMessageContext,
        request: OpenConversationReq,
    ): OpenConversationRes {
        if (actor.role != SupportChatRoles.Customer) {
            throw IllegalStateException("Only customer actors can open a conversation.")
        }

        // --8<-- [start:doc-sc-open-actor]
        val opened =
            entrySpot
                .context()
                .outbound()
                .requestToChannel(
                    SampleNames.ApiChannel,
                    OpenConversationApiReq(actor.participantId, actor.displayName, request.subject),
                )
                .timeout(SampleTimings.RequestTimeout)
                .submit(OpenConversationApiRes::class.java)
                .await()

        val joined =
            actor.scheduleConversationJoin(
                opened.conversationId,
                request.subject,
                JoinConversationReq(actor.participantId, actor.role, actor.displayName),
            )
        return OpenConversationRes(opened.conversationId, joined.state)
        // --8<-- [end:doc-sc-open-actor]
    }
}
